### PR TITLE
Check url

### PR DESCRIPTION
**Pull request by seanmtracey**
Checks if url is likely to be viewable in Screens

### CODE REVIEW

File: client/admin/js/main.js [hunk 1]
```diff
@@ -13,6 +13,28 @@ const HierarchicalNav = require('o-hierarchical-nav');
...
+const troubleURLS = [];
+
+function pointOutTroubleMakers(){
+	const activeLinks = Array.from(document.querySelectorAll('.screen-page'));
+
+	activeLinks.forEach(function(activeLink){
+
+		const link = activeLink.getAttribute('href');
+
+		debugger;
```
Comment (JakeChampion) [inline client/admin/js/main.js:25]: We should remove this

File: client/admin/js/main.js [hunk 2]
```diff
@@ -13,6 +13,28 @@ const HierarchicalNav = require('o-hierarchical-nav');
...
+
+		troubleURLS.forEach(function(url){
```
Comment (JakeChampion) [inline client/admin/js/main.js:27]: We could replace this with the [includes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes) method.

Example:

```
if (troubleURLS.includes(link)) {
  activeLink.setAttribute('data-troublsome-url', 'true');
}
```

I'm not 100% but I think we could remove the `if` statement and go with:
`activeLink.setAttribute('data-troublsome-url', troubleURLS.includes(link));`

File: client/admin/js/main.js [hunk 3]
```diff
@@ -130,7 +153,20 @@ window.screensInit = function() {
...
 			}
-			api('addUrl', data);
+
+			api('addUrl', data)
+				.then(function(res){
+					const canBeViewed = res.viewable;
+
+					if(!canBeViewed){
+
+						if(troubleURLS.indexOf(data.url) === -1){
```
Comment (JakeChampion) [inline client/admin/js/main.js:163]: We could replace this with the [includes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes) method.

Example:

```
if (!troubleURLS.includes(data.url)) {
  troubleURLS.push(data.url);
}
```

File: package.json [hunk 4]
```diff
@@ -19,11 +19,13 @@
     "express": "~4.12.2",
     "express-ftwebservice": "^2.0.2",
     "express-handlebars": "^2.0.1",
+    "fetch-er": "0.0.10",
```
Comment (JakeChampion) [inline package.json:22]: I don't see this being used in the codebase, could we remove it?

File: package.json [hunk 5]
```diff
@@ -19,11 +19,13 @@
...
     "gulp": "latest",
     "htmlparser": "^1.7.7",
     "image-type": "^2.1.0",
     "is-url-superb": "^2.0.0",
     "lodash": "^3.10.1",
+    "lodash.curry": "^3.0.2",
```
Comment (JakeChampion) [inline package.json:28]: We have the whole of lodash being pulled in as a dependency on the line above, no need for this extra dependency :-)

File: server/routes/api.js [hunk 6]
```diff
@@ -5,11 +5,54 @@ const router = require('express').Router(); // eslint-disable-line new-cap
...
+function createJsonResponse (res, statusCode, success, data){
+  res.status(statusCode)
+  .json({
+    success,
+    data,
+    statusCode: code
+  });
+}
+
+const response = curry(createJsonResponse);
```
Comment (JakeChampion) [inline server/routes/api.js:25]: This isn't being used anywhere

File: server/routes/api.js [hunk 7]
```diff
@@ -195,4 +247,40 @@ router.post('/reload', function(req, res) {
 	res.json(true);
 });
 
+router.post('/is-viewable', function(req, res){
+
+	const url = req.body.url;
+
+	fetch(url, {
```
Comment (JakeChampion) [inline server/routes/api.js:254]: `fetch` is not defined in this module

File: server/routes/api.js [hunk 8]
```diff
@@ -195,4 +247,40 @@ router.post('/reload', function(req, res) {
...
+			method: 'head'
+		})
+		.then(function(response){
+			return response.headers.get('x-frame-options');
```
Comment (JakeChampion) [inline server/routes/api.js:258]: I think this is a synchronous function, there's no need to have it go into another thenable as that is actually less performant.

File: server/routes/api.js [hunk 9]
```diff
@@ -195,4 +247,40 @@ router.post('/reload', function(req, res) {
...
+		})
+		.then(function(xFrameHeader){
+
+			if(xFrameHeader === null){
+				res.json({
```
Comment (JakeChampion) [inline server/routes/api.js:263]: Could this make use of the createResponse function that is defined at the top of the file?

File: client/admin/js/main.js (hunk 1 above)
Comment (seanmtracey) [inline client/admin/js/main.js:25]: ~~Absolutely~~ I agreed with this too quickly. This is how we add the styles to indicate that the URL may be at troublesome. We can't do it on the API request, as the list isn't updated until we trigger the updateScreens function (with a WS message). Is there a better way to do it that I'm missing (other than your suggestions below)?

File: package.json (hunk 4 above)
Comment (seanmtracey) [inline package.json:22]: Yeah, this is a hangover from my debugging. Will remove

File: package.json (hunk 5 above)
Comment (seanmtracey) [inline package.json:28]: This is what happens when you --save-dev without thinking. My bad!

File: server/routes/api.js (hunk 8 above)
Comment (seanmtracey) [inline server/routes/api.js:258]: I'll check and adjust accordingly

File: server/routes/api.js (hunk 9 above)
Comment (seanmtracey) [inline server/routes/api.js:263]: Yup.

File: server/routes/api.js (hunk 6 above)
Comment (seanmtracey) [inline server/routes/api.js:25]: It's being used on line 31. Adding node-fetch back in too as it's used for the URL shortening code.

File: client/admin/js/main.js (hunk 2 above)
Comment (seanmtracey) [inline client/admin/js/main.js:27]: I think this may be the case. Will give it a go and see how it behaves.

File: client/admin/js/main.js [hunk 10]
```diff
@@ -13,6 +13,25 @@ const HierarchicalNav = require('o-hierarchical-nav');
...
+
+function pointOutTroubleMakers(){
+	const activeLinks = Array.from(document.querySelectorAll('.screen-page'));
+
+	activeLinks.forEach(function(activeLink){
+
+		const link = activeLink.getAttribute('href');
+
+		troubleURLS.forEach(function(url){
+			if(url === link){
```
Comment (JakeChampion) [inline client/admin/js/main.js:28]: Could we use array includes here? It might need to be required in as a module

File: client/admin/js/main.js [hunk 11]
```diff
@@ -130,7 +150,18 @@ window.screensInit = function() {
 			if ($('#date').val() || $('#time').val()) {
 				data.dateTimeSchedule = moment(`${$('#date').val() || moment().format('YYYY-MM-DD')} ${$('#time').val() || moment().format('HH:mm')}`).valueOf();
 			}
-			api('addUrl', data);
+
+			api('addUrl', data)
+				.then(function(res){
+					const canBeViewed = res.data.viewable;
+					if(!canBeViewed){
+						if(troubleURLS.indexOf(data.url) === -1){
```
Comment (JakeChampion) [inline client/admin/js/main.js:163]: Same here for array includes method

File: server/routes/api.js [hunk 12]
```diff
@@ -113,8 +154,15 @@ router.post('/addUrl', function(req, res, next) {
...
 
-
-			res.json(true);
+			return checkIsViewable(url)
+				.then(isViewable => {
+					response(res, 200, true, { viewable : isViewable })
+				})
+				.catch(err => {
+					console.log(err);
+					res.json(true);
```
Comment (JakeChampion) [inline server/routes/api.js:167]: This should use the response function

**Comment by seanmtracey** [comment]
No Hover
no_hover
::hover
on_hover

**Pull request merged**